### PR TITLE
Further Improve reporting

### PR DESCRIPTION
**Pull request by amrohassaan**
examples: build report & template: Add missing testrun objects to the
report to build accurate hyperlinks to tests view. Add missing job_url
to LAVA job. Add missing metrics. Misc fixes to template (bg color for
status & proper aligning to suite names)

### CODE REVIEW

File: examples/build_report.py [hunk 1]
```diff
@@ -55,21 +55,33 @@
...
-    # group tests by suite
-    for test in tests.values():
-        suite_name = squad_suites[test.suite].slug
-        if suite_name in suites.keys():
-            suites[suite_name].append(test)
-        else:
-            suites.update({suite_name: [test]})
+    metric_suites = {}
+    for testrun in testruns.values():
+        envs_tr_combo.update({(env.slug, testrun): (suites, metric_suites)})
```
Comment (mwasilew) [inline examples/build_report.py]: I don't like this. You're giving suites in random order which looks ugly in the final report. Also using tuple as a key is asking for trouble.

File: examples/build_report_template.html [hunk 2]
```diff
@@ -69,29 +69,48 @@ <h2>Details for <a href="{{ base_url + 'build/' + build.version + '/' }}">build
...
-                    {% for suite_name, results in env_results.items() -%}
+                    {% for suite_name, results in env_results[0].items() -%}
                     <tr>
                         <td rowspan="{{ results|length + 1}}">{{ suite_name }}</td>
                     </tr>
                     {% for test in results -%}
                     <tr>
                         <td>
-                            <a href="{{ base_url + 'tests/' + test.name }}">{{ test.short_name }}</a>
+                            <a href="{{ base_url + 'build/' + build.version + '/testrun/' + envs_tr_tuple[1].job_id + '/suite/' + suite_name + '/tests/' }}">{{ test.short_name }}</a>
```
Comment (mwasilew) [inline examples/build_report_template.html]: did you try it? Links are totally incorrect.

File: examples/build_report.py [hunk 3]
```diff
@@ -55,21 +55,33 @@
...
+            suite_name = squad_suites[test.suite].slug
+            if suite_name in suites.keys():
+                suites[suite_name].append(test)
+            else:
+                suites.update({suite_name: [test]})
+
+        metrics = testrun.metrics()
+        if metrics:
+            for metric in metrics.values():
+                suite_name, metric_name = metric.name.split("/")
```
Comment (mwasilew) [inline examples/build_report.py]: this is wrong and should be fixed in the API.

File: examples/build_report_template.html [hunk 4]
```diff
@@ -69,29 +69,48 @@ <h2>Details for <a href="{{ base_url + 'build/' + build.version + '/' }}">build
             </div>
             <div id="testrun-list">
                 <h2>LAVA test results</h2>
-                {% for env_name, env_results in environments.items() %}
-                <h3>{{ env_name }}</h3>
+                {% for envs_tr_tuple, env_results in envs_tr_combo.items() %}
+                <h3>{{ envs_tr_tuple[0] }}</h3>
                 <table class="table" border="1">
+                    <tr>
```
Comment (mwasilew) [inline examples/build_report_template.html]: If at all, this should be outside of the table. Also giving just the test run ID seems a bad idea.

File: examples/build_report_template.html [hunk 5]
```diff
@@ -69,29 +69,48 @@ <h2>Details for <a href="{{ base_url + 'build/' + build.version + '/' }}">build
...
                     {% endfor %}
                     {% endfor %}
+                    {% for suite_name, results in env_results[1].items() -%}
+                    <tr>
+                        <td rowspan="{{ results|length + 1}}">{{ suite_name }}</td>
+                    </tr>
+                    {% for metric in results -%}
+                    <tr>
+                        <td>
+                            {{ metric.name.split("/")[1]}}
```
Comment (mwasilew) [inline examples/build_report_template.html]: no link?

**Comment by mwasilew** [comment]
This is very wrong. It produces wrong report with incorrect links and is inconsistent.

**Comment by amrohassaan** [comment]
@mwasilew fixed now.

**Comment by chaws** [comment]
It's looking better now. I'm attaching a sample pdf report generated using this PR:  [build_50_report.pdf](https://github.com/Linaro/squad-client/files/4354692/mozilla.pdf)

Now I'm going to open another PR that changes the intro text to be jinja2-compatible.

**Comment by mwasilew** [comment]
I still don't like the breakdown by testjob. IMHO it makes no sense. Testjob/testrun should only be used to create links.

**Comment by chaws** [comment]
@mwasilew @amrohassaan  I thought about creating under Linaro's github organization. The repo structure would be something like:
```
Linaro/squad-reports
  - schneider
  - lkft
  - ...
```

Then we can make the generic `build_report` in squad-client to abstract testrun/testjob as Milosz suggested. In the squa-reports repository, under schneider folder, we'd keep the template of this PR.

What do you guys think of that?

**Comment by amrohassaan** [comment]
I am closing this now as obsolete in favour of [this PR](https://github.com/Linaro/squad-client/pull/23)

**Pull request closed without merging**